fem/pfespace.cpp

```diff
@@ -1038,263 +1070,635 @@ void ParFiniteElementSpace::ConstructTrueNURBSDofs()
+         return;
+      }
+      dof -= ngvdofs;
+      if (dof < ngedofs) // ghost edge
+      {
+         int ne = fec->DofForGeometry(Geometry::SEGMENT);
+         entity = 1, index = pncmesh->GetNEdges() + dof / ne, edof = dof % ne;
+         return;
       }
-      if (ne)
+      dof -= ngedofs;
+      if (dof < ngfdofs) // ghost face
       {
-         for (i = 0; i < 4; i++)
+         int nf = fec->DofForGeometry(mesh->GetFaceBaseGeometry(0));
+         entity = 2, index = pncmesh->GetNFaces() + dof / nf, edof = dof % nf;
+         return;
+      }
+      MFEM_ABORT("Out of range DOF.");
+   }
+}
+
+/** Represents an element of the P matrix. The column number is global and
+ *  corresponds to vector dimension 0. The other dimension columns are offset
+ *  by 'stride'.
+ */
+struct PMatrixElement
+{
+   HYPRE_Int column, stride;
+   double value;
+
+   PMatrixElement(HYPRE_Int col = 0, HYPRE_Int str = 0, double val = 0)
+      : column(col), stride(str), value(val) {}
+
+   bool operator<(const PMatrixElement &other) const
+   { return column < other.column; }
+
+   typedef std::vector<PMatrixElement> List;
+};
+
+/** Represents one row of the P matrix, for the construction code below.
+ *  The row is complete: diagonal and offdiagonal elements are not distinguished.
+ */
+struct PMatrixRow
+{
+   PMatrixElement::List elems;
+
+   /// Add other row, times 'coef'.
+   void AddRow(const PMatrixRow &other, double coef)
+   {
+      elems.reserve(elems.size() + other.elems.size());
+      for (unsigned i = 0; i < other.elems.size(); i++)
+      {
+         const PMatrixElement &oei = other.elems[i];
+         elems.push_back(
+            PMatrixElement(oei.column, oei.stride, coef * oei.value));
+      }
+   }
+
+   /// Remove duplicate columns and sum their values.
+   void Collapse()
+   {
+      if (!elems.size()) { return; }
+      std::sort(elems.begin(), elems.end());
+
+      int j = 0;
+      for (unsigned i = 1; i < elems.size(); i++)
+      {
+         if (elems[j].column == elems[i].column)
          {
-            if (corner[i] && corner[(i+1) % 4])
-            {
-               for (k = 0; k < ne; k++)
-               {
-                  slave_dofs[4*nv + i*ne + k] = INVALID_DOF;
-               }
-            }
+            elems[j].value += elems[i].value;
+         }
+         else
+         {
+            elems[++j] = elems[i];
          }
       }
+      elems.resize(j+1);
    }
-}
 
-void ParFiniteElementSpace
-::AddSlaveDependencies(DepList deps[], int master_rank,
-                       const Array<int> &master_dofs, int master_ndofs,
-                       const Array<int> &slave_dofs, DenseMatrix& I) const
+   void write(std::ostream &os, double sign) const
+   {
+      mfem::write<int>(os, elems.size());
+      for (unsigned i = 0; i < elems.size(); i++)
+      {
+         const PMatrixElement &e = elems[i];
+         mfem::write<HYPRE_Int>(os, e.column);
+         mfem::write<int>(os, e.stride);
```

Comment:
If `stride` is removed, it won't matter but its type here is wrong. Unless it is truncated on purpose?

Reply:
This is on purpose to save message size (stride is either 1 or ndofs, so 32-bits are OK). PMatrixElement::stride could really be an int but we would save nothing due to struct padding, so I left it the same type as 'column'...